File: LeetcodePython/LE002AddTwoNumbers.py
from utility.ListNode import ListNode


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        inc = False
        dummy = ListNode(0)
        node = dummy
        while l1 or l2 or inc:
            tmp = 0

            if l1:
                tmp += l1.val
                l1 = l1.next
            if l2:
                tmp += l2.val
                l2 = l2.next

            if inc:
                tmp += 1
                inc = False
            if tmp >= 10:
                tmp %= 10
                inc = True
            node.next = ListNode(tmp)
            node = node.next
        return dummy.next

l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l1.next = l2
l2.next = l3
l3.next = l4
l11 = ListNode(1)
l12 = ListNode(2)
l13 = ListNode(3)
l14 = ListNode(6)
l11.next = l12
l12.next = l13
l13.next = l14


if __name__ == '__main__':
    test = Solution()
    l = test.addTwoNumbers(l1, l11)
    l.print()

File: LeetcodeJava/src/main/java/org/khwu/LE002AddTwoNumbers.java
package org.khwu;
import org.khwu.utility.ListNode;

public class LE002AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    	ListNode head = new ListNode(0);
    	ListNode re = head;
    	int val;
    	boolean inc = false;
    	while(l1 != null || l2 != null || inc){
    		val = inc ? 1 : 0; 
    		if(l1 != null){
    			val += l1.val;
    			l1 = l1.next;
    		}
    		if(l2 != null) {
    			val += l2.val;
    			l2 = l2.next;
    		}
    		inc = val >= 10 ? true : false;
    		re.next = new ListNode(val % 10);
//    		System.out.println(re.val);
    		re = re.next;
    	}
    	re = null;
    	return head.next;
    }
//    public ListNode helpAddNode(ListNode l1, ListNode l2, boolean inc){
//    	ListNode re;
//    	int val = inc ? 1 : 0; 
//		if(l1 != null) val += l1.val;
//		if(l2 != null) val += l2.val;
//		re = new ListNode(val);
//    	}else{
//    		if(l1 == null && l2 == null) re = null;
//    		else if(l1 == null) re = l2.next;
//        	else if(l2 == null) re = l1.next;
//        	else{
//        		re = new ListNode((l1.val + l2.val) % 10);
//        		if (l1.val + l2.val >=10) re.next = helpAddNode(l1.next, l2.next, true);
//        		else re.next = helpAddNode(l1.next, l2.next, false);
//        	}
//    	}
//    	return re;
//    }
    
    public static void main(String[] args){
    	ListNode L1 = new ListNode(2);
		ListNode L2 = new ListNode(4);
		ListNode L3 = new ListNode(3);
		ListNode L4 = new ListNode(5);
		ListNode L5 = new ListNode(6);
		ListNode L6 = new ListNode(7);
		L1.next = L2;
		L2.next = L3;
		L4.next = L5;
		L5.next = L6;
		LE002AddTwoNumbers test = new LE002AddTwoNumbers();
		ListNode tt = test.addTwoNumbers(L1, L4);
		while(tt != null){
			System.out.print(tt.val + " ");
			tt = tt.next;
		}
		
    }
}
